### PR TITLE
[Fix] Ensure we set the eval/train flag correctly on accelerator model 

### LINKED ISSUE #6876: Latest FairScale + Sharded Training crashes using default trainer parameters
## 🐛 Bug

When validation/training is used (as default with the boring model) sharded crashes. This is because internally SDP relies on knowing the training state of the model, and when we run the validation sanity check, we do not set the eval mode correctly on the SDP model itself, so it waits for grads to be reduced since the module is in `train` mode.

```python
import os

import torch
from torch.utils.data import Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    """
    >>> RandomDataset(size=10, length=20)  # doctest: +ELLIPSIS
    <...bug_report_model.RandomDataset object at ...>
    """

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    """
    >>> BoringModel()  # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
    BoringModel(
      (layer): Linear(...)
    )
    """

    def __init__(self):
        """
        Testing PL Module

        Use as follows:
        - subclass
        - modify the behavior for what you want

        class TestModel(BaseTestModel):
            def training_step(...):
                # do your own thing

        or:

        model = BaseTestModel()
        model.training_epoch_end = None

        """
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def step(self, x):
        x = self.layer(x)
        out = torch.nn.functional.mse_loss(x, torch.ones_like(x))
        return out

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"x": loss}

    def test_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"y": loss}

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


def test_run():
    # fake data
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    test_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    # model
    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        max_epochs=1,
        plugins='ddp_sharded',
        gpus=1,
        weights_summary=None,
    )
    trainer.fit(model, train_data, val_data)
    trainer.test(test_dataloaders=test_data)


if __name__ == '__main__':
    test_run()
```

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Fixes #6876.

This PR is ready, with more information underneath.

We also require https://github.com/facebookresearch/fairscale/pull/587 to be merged and included in next release, but since we do not rely on the upstream release yet, this fix isn't going to change any FairScale related tests on CI. This will probably also block #6152 as we move from our own fork to the latest FairScale release.

In terms of testing, once #6152 is merged to deprecate the current Pipe implementation, this fix will come to fruition as the FairScale version will be reliant on the upstream release.

Overall however, I notice in many places we call `train` `eval` on just the lightning module, not the accelerator wrapped model. In the case of custom implementations like SDP, this is an issue. I swapped this to using `self.model` which should recursively set the lightning module when it is wrapped.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]
great catch!!! let's add comments here for when to use self.model vs when to directly use self.lightning_module

**Comment by SeanNaren** [comment]
> great catch!!! let's add comments here for when to use self.model vs when to directly use `self.lightning_module`

Thanks @ananthsub!

The difference from the names can be quite confusing, and in most cases the choice is clearer; if you want to access `lightning_module` internals/functions, use `self.trainer.lightning_module`. And eventually for FSDP or where you'd like to access the wrapped model, you can use `self.trainer.model` or `self.model` in the trainer :)

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -665,6 +665,8 @@ def run_evaluation(self, on_epoch=False):
         # ref model
         model = self.lightning_module
         model.zero_grad()
+
+        self.model.eval()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:669]: careful, right above here we have a special hook `self.evaluation_loop.on_evaluation_model_eval`
we probably don't want to override this behaviour if the user implements this hook
same for the other places (e.g. predict)
confidence on this statement: ~60%

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -786,6 +788,7 @@ def run_predict(self):
 
         # enable eval mode + no grads
         self.predict_loop.on_predict_model_eval()
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:790]: Same here.

**Comment by SeanNaren** [comment]
Ah great point, forgot `lightning_module.eval()` and `lightning_module.train()` are hooks. What's advised here? I could change the default to `self.trainer.model.eval()` and `self.trainer.model.train()` within the hooks or I could add this additional logic outside the hook. cc @awaelchli @tchaton

**Comment by tchaton** [comment]
> Ah great point, forgot `lightning_module.eval()` and `lightning_module.train()` are hooks. What's advised here? I could change the default to `self.trainer.model.eval()` and `self.trainer.model.train()` within the hooks or I could add this additional logic outside the hook. cc @awaelchli @tchaton

Sounds good to me !

**Comment by SeanNaren** [comment]
Thanks @tchaton just a slight mistake I should be referring to the module hooks `on_test_model_train` `on_test_model_eval` `on_predict_model_eval` that will change.

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -786,6 +786,7 @@ def run_predict(self):
...
+
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/trainer.py:789]:
```suggestion
```

**Review by awaelchli — APPROVED** [review]
looks good <3

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Comment by carmocca** [comment]
@SeanNaren the test is failing

**Comment by SeanNaren** [comment]
The test showed that we call the predict hook twice, I've remedied this and it now follows the same behaviour as run_evaluation and run_train

**Review by ananthsub — APPROVED** [review]
